### PR TITLE
Expose alpha_y, beta_y parameters

### PR DESCRIPTION
**Pull request by cmower**
Exposes `alpha_y` and `beta_y` to all DMP constructors and enables vectorization while maintaining backward compatibility with scalar values. 

Addresses #57.

**Are there unit tests?**

Unit tests are updated to handle new features. Pytest going green.

**Does it have docstrings?**

Docstrings have been updated where appropriate.

**Is it included in the API documentation?**

NA

**Run flake8 and pylint**

Done

**Should it be part of the readme?**

No

**Should it be included in any example script?**

Not necessarily.

### CODE REVIEW

File: test/test_cartesian_dmp.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def test_compare_python_cython():
         current_y=np.array([1.0, 0.0, 0.0, 0.0]), current_yd=np.zeros([3]),
         goal_y=np.array([1.0, 0.0, 0.0, 0.0]), goal_yd=np.zeros([3]), goal_ydd=np.zeros([3]),
         start_y=np.array([1.0, 0.0, 0.0, 0.0]), start_yd=np.zeros([3]), start_ydd=np.zeros([3]),
-        goal_t=2.0, start_t=0.0, alpha_y=25.0, beta_y=6.25,
+        goal_t=2.0, start_t=0.0, alpha_y=25.0*np.ones(3), beta_y=6.25*np.ones(3),
```
Comment (AlexanderFabisch) [inline test/test_cartesian_dmp.py:78]: PEP8 suggests to add spaces around operators

File: movement_primitives/dmp/_dmp.py [hunk 2]
```diff
@@ -255,10 +255,10 @@ def dmp_step_euler(
     start_t : float
         Time at the start.
 
-    alpha_y : float
+    alpha_y : array, shape (n_dims,)
```
Comment (AlexanderFabisch) [inline movement_primitives/dmp/_dmp.py:258]: * It seems like float is still a valid option and will be "broadcasted". This should be documented.
* The default value should also be documented.

File: movement_primitives/dmp/_dmp.py [hunk 3]
```diff
@@ -255,10 +255,10 @@ def dmp_step_euler(
...
         Constant in transformation system.
 
-    beta_y : float
+    beta_y : array, shape (n_dims,)
```
Comment (AlexanderFabisch) [inline movement_primitives/dmp/_dmp.py:261]: Same here.

File: movement_primitives/dmp/_dmp.py [hunk 4]
```diff
@@ -380,6 +380,12 @@ class DMP(WeightParametersMixin, DMPBase):
         is changed and the trajectory is scaled by interpolating between
         the old and new scaling of the trajectory.
 
+    alpha_y : float, list with length n_dims, or array with shape (n_dims,), optional (default: 25.0)
```
Comment (AlexanderFabisch) [inline movement_primitives/dmp/_dmp.py:383]: I would just write `array-like, shape (n_dims,)`

File: movement_primitives/dmp/_dmp.py [hunk 5]
```diff
@@ -416,8 +422,31 @@ def __init__(self, n_dims, execution_time=1.0, dt=0.01,
 
         self._init_forcing_term()
 
-        self.alpha_y = 25.0
-        self.beta_y = self.alpha_y / 4.0
+        if isinstance(alpha_y, float):
```
Comment (AlexanderFabisch) [inline movement_primitives/dmp/_dmp.py:425]: * This whole thing looks a bit fragile. What if the input is a jax.numpy.ndarray or a tuple? What if the input is an integer (e.g., `25`)? I would like to avoid manual `isinstance` checks and `type` calls.
    * `isinstance(..., float)` should be replaced by checking for a scalar numerical value and cast it to a 1d numpy array of dtype float
    * `isinstance(..., (np.ndarray, list))` should be replaced by checking for an iterable or array-like interface and cast it to a 1d numpy array of dtype float
* I would also suggest to extract this as a more general function that will be applied to `alpha_y` and `beta_y`. 

`sklearn` has a similar function: [`sklearn.utils.check_array`](https://github.com/scikit-learn/scikit-learn/blob/8a2d5ffa0/sklearn/utils/validation.py#L718). Maybe we can use this for inspiration (although it is certainly way to complex for our use case).

File: movement_primitives/dmp/_dmp_with_final_velocity.py [hunk 6]
```diff
@@ -35,6 +35,12 @@ class DMPWithFinalVelocity(WeightParametersMixin, DMPBase):
         Gain for proportional controller of DMP tracking error.
         The domain is [0, execution_time**2/dt].
 
+    alpha_y : float, list with length n_dims, or array with shape (n_dims,), optional (default: 25.0)
```
Comment (AlexanderFabisch) [inline movement_primitives/dmp/_dmp_with_final_velocity.py:38]: same applies here

File: movement_primitives/dmp/_dmp_with_final_velocity.py [hunk 7]
```diff
@@ -63,8 +70,23 @@ def __init__(self, n_dims, execution_time=1.0, dt=0.01,
 
         self._init_forcing_term()
 
-        self.alpha_y = 25.0
-        self.beta_y = self.alpha_y / 4.0
+        if isinstance(alpha_y, float):
```
Comment (AlexanderFabisch) [inline movement_primitives/dmp/_dmp_with_final_velocity.py:73]: this is duplicate code

File: movement_primitives/dmp/_dual_cartesian_dmp.py [hunk 8]
```diff
@@ -239,8 +245,29 @@ def __init__(self, execution_time=1.0, dt=0.01, n_weights_per_dim=10,
 
         self._init_forcing_term()
 
-        self.alpha_y = 25.0
-        self.beta_y = self.alpha_y / 4.0
+        if isinstance(alpha_y, float):
```
Comment (AlexanderFabisch) [inline movement_primitives/dmp/_dual_cartesian_dmp.py:248]: duplicate code

File: movement_primitives/dmp/_state_following_dmp.py [hunk 9]
```diff
@@ -34,8 +34,26 @@ def __init__(self, n_dims, execution_time, dt=0.01, n_viapoints=10,
 
         alpha_z = canonical_system_alpha(0.01, self.execution_time, 0.0)
 
-        self.alpha_y = 25.0
-        self.beta_y = self.alpha_y / 4.0
+        if isinstance(alpha_y, float):
```
Comment (AlexanderFabisch) [inline movement_primitives/dmp/_state_following_dmp.py:37]: duplicate code

**Comment by AlexanderFabisch** [comment]
Thanks a lot for the contribution. I commented on some parts.

I think an example script would actually be interesting! What is the use case that you are aiming for? Is there a simple way to demonstrate that different values for alpha and beta make sense?

Btw are you aiming for combinations of alpha and beta that are not critically damped? If not, then we can also set beta based on the value of alpha or vice versa.

File: test/test_cartesian_dmp.py (hunk 1 above)
Comment (cmower) [reply test/test_cartesian_dmp.py]: Addressed in [0584376](https://github.com/dfki-ric/movement_primitives/pull/59/commits/0584376262360124530b4f2e5704ca6fd314496b)

File: movement_primitives/dmp/_dmp.py (hunk 4 above)
Comment (cmower) [reply movement_primitives/dmp/_dmp.py]: Addressed in [7e5d15e](https://github.com/dfki-ric/movement_primitives/pull/59/commits/7e5d15ea3cd1587d8f2bbca949f1a8385325a8f8)

File: movement_primitives/dmp/_dmp_with_final_velocity.py (hunk 6 above)
Comment (cmower) [reply movement_primitives/dmp/_dmp_with_final_velocity.py]: Addressed in [7e5d15e](https://github.com/dfki-ric/movement_primitives/pull/59/commits/7e5d15ea3cd1587d8f2bbca949f1a8385325a8f8)

File: movement_primitives/dmp/_dmp.py (hunk 5 above)
Comment (cmower) [reply movement_primitives/dmp/_dmp.py]: Please see [9103981](https://github.com/dfki-ric/movement_primitives/pull/59/commits/91039818c8624e4c282220596192a4bb0863932a), and let me know if this works for you?

File: movement_primitives/dmp/_dmp_with_final_velocity.py (hunk 7 above)
Comment (cmower) [reply movement_primitives/dmp/_dmp_with_final_velocity.py]: Please see [9103981](https://github.com/dfki-ric/movement_primitives/pull/59/commits/91039818c8624e4c282220596192a4bb0863932a), and let me know if this works for you?

File: movement_primitives/dmp/_dual_cartesian_dmp.py (hunk 8 above)
Comment (cmower) [reply movement_primitives/dmp/_dual_cartesian_dmp.py]: Please see [9103981](https://github.com/dfki-ric/movement_primitives/pull/59/commits/91039818c8624e4c282220596192a4bb0863932a), and let me know if this works for you?

File: movement_primitives/dmp/_state_following_dmp.py (hunk 9 above)
Comment (cmower) [reply movement_primitives/dmp/_state_following_dmp.py]: Please see [9103981](https://github.com/dfki-ric/movement_primitives/pull/59/commits/91039818c8624e4c282220596192a4bb0863932a), and let me know if this works for you?

**Comment by cmower** [comment]
> Thanks a lot for the contribution. I commented on some parts.
> 

Hi @AlexanderFabisch, no problem, thanks for the comments. I've addressed all the above, please let me know if you would like any further changes?

> I think an example script would actually be interesting! What is the use case that you are aiming for? Is there a simple way to demonstrate that different values for alpha and beta make sense?
> 

Sure, I will make a small example. Currently, I am just learning some simple demonstrations via kinaesthetic teaching on a fixed-base robot arm - later perhaps something more interesting. 

> Btw are you aiming for combinations of alpha and beta that are not critically damped? If not, then we can also set beta based on the value of alpha or vice versa.

If I understand correct, I think this would add more complexity to the code/documentation than needed - in my opinion, best to leave this choice to the user and set what they want. Please let me know your thoughts?

**Comment by cmower** [comment]
Hi @AlexanderFabisch, added an example. Please let me know if you would like any other changes? Otherwise, I think this is good to merge?

File: movement_primitives/dmp/_dmp.py (hunk 5 above)
Comment (AlexanderFabisch) [reply movement_primitives/dmp/_dmp.py]: I'd suggest to extract it as a function and give it a more generic name. It's not specific to DMPs.

File: examples/plot_dmp_2d_gain.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
+"""
+==================
+Effect of DMP Gain
+==================
+
+Demonstrates how modifying DMP gains (alpha_y, beta_y) affects 
+the resulting trajectory reproduction.
+
```
Comment (AlexanderFabisch) [inline examples/plot_dmp_2d_gain.py:8]: empty lines should be removed

File: examples/plot_dmp_2d_gain.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+"""
+
+print(__doc__)
+
+import matplotlib.pyplot as plt
+import numpy as np
+from movement_primitives.dmp import DMP
+
```
Comment (AlexanderFabisch) [inline examples/plot_dmp_2d_gain.py:17]: PEP8: add one more blank line

File: movement_primitives/dmp/_dmp.py (hunk 5 above)
Comment (cmower) [reply movement_primitives/dmp/_dmp.py]: I will change the method name, but any suggestions where to put `_process_gain_input`? Either `movement_primitives/base.py` or I could create `movement_primitives/utils.py` perhaps?

File: examples/plot_dmp_2d_gain.py (hunk 10 above)
Comment (cmower) [reply examples/plot_dmp_2d_gain.py]: Addressed in [f490348](https://github.com/dfki-ric/movement_primitives/pull/59/commits/f4903481e2804a15fb624d28cd8faa764906c5bf)

File: movement_primitives/utils.py [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import numpy as np
+
+
+def ensure_1d_array(value, dim, label):
```
Comment (AlexanderFabisch) [inline movement_primitives/utils.py:4]: * For consistency, could you rename `dim` to `n_dims`?
* ... and label reminds me of classification, maybe something like `var_name`?

**Comment by AlexanderFabisch** [comment]
I modified `ensure_1d_array` (unit tests, documentation, bugfixe) in d187c1598c8a429df4c546b2a98994c0a468b80d and merged this PR.

Thanks for the contribution!
